### PR TITLE
Added a feature to automate unsealing operation of vault

### PR DESCRIPTION
**Pull request by Shifna12Zarnaz**
Added a file vault2.go under server/pkg/client/vault2.go.It ll generate unseal keys and store that keys in vault and check the status of vault whether it is sealed or not for every 1 min and automatically unseal it if vault is sealed.Tested separately as a package.Also in tha server main.go added the go routine to the monitoring function

### CODE REVIEW

File: server/pkg/client/vault2.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		_, err := v.client.Sys().Unseal(key)
+		if err != nil {
+			flag = false
+			fmt.Println("Error while unsealing", err)
+			return err
+		}
+
+	}
+	if flag {
+		fmt.Println("Unsealed")
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:61]: what is the use of this print statement

File: server/pkg/client/vault.go [hunk 2]
```diff
@@ -38,6 +39,7 @@ func NewVault() (*Vault, error) {
 
 	fmt.Println("token is ", token)
 	client.SetToken(token)
+	fmt.Println("Client is", client)
```
Comment (srikrishnabh1) [inline server/pkg/client/vault.go:42]: what is the use if this print statement

File: server/pkg/client/vault.go [hunk 3]
```diff
@@ -84,3 +86,41 @@ func (v *Vault) GetCert(secretName, customerID string) (map[string]string, error
...
+	}
+	secretData := map[string]interface{}{
+		"username": username,
+		"password": password,
+	}
+
+	ctx := context.Background()
+
+	// Write the secret
+	_, err := v.client.KVv2(secretName).Put(ctx, astradbcreds, secretData)
```
Comment (srikrishnabh1) [inline server/pkg/client/vault.go:101]: post credentials is generic function name, why is it astradb specific names passed

File: server/pkg/client/vault.go [hunk 4]
```diff
@@ -84,3 +86,41 @@ func (v *Vault) GetCert(secretName, customerID string) (map[string]string, error
...
+	//fmt.Println("")
```
Comment (srikrishnabh1) [inline server/pkg/client/vault.go:102]: remove dead code

File: server/pkg/client/vault.go [hunk 5]
```diff
@@ -84,3 +86,41 @@ func (v *Vault) GetCert(secretName, customerID string) (map[string]string, error
...
+	if err != nil {
+		return fmt.Errorf("failed to write secret: %w", err)
+	}
+
+	log.Println("Secret written successfully.")
```
Comment (srikrishnabh1) [inline server/pkg/client/vault.go:107]: Can you make it debug

File: server/pkg/client/vault.go [hunk 6]
```diff
@@ -84,3 +86,41 @@ func (v *Vault) GetCert(secretName, customerID string) (map[string]string, error
...
+	return nil
+}
+
+func (v *Vault) GetCredentials(secretName, astradbcreds string) (string, string, error) {
```
Comment (srikrishnabh1) [inline server/pkg/client/vault.go:111]: we should use generic name the variable name is specific to astradb

File: server/pkg/client/vault2.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
+package client
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:1]: Refactor the code and make it modular.

File: server/pkg/client/vault2.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	if err != nil {
+		fmt.Println("Error while creating clientset", err)
+	}
+
+	namespace := "default"       // Namespace where you want to create the Secret
+	secretName := "vault-secret" // Name of the Secret
+
+	// Retrieve the secret with the given name
+	secret, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:89]: bad practice

File: server/pkg/client/vault2.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+		// Secret not found, create a new one
+		if err.Error() == "secrets \""+secretName+"\" not found" {
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:91]: we should not check error like this

File: server/pkg/client/vault2.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+			values = append(values, keys)
+			fmt.Println("Key is ", keys)
+		}
+
+		fmt.Printf("Secret '%s' found in namespace '%s'\n", secret.Name, secret.Namespace)
+		// Use the secret as needed
+	}
+	return values
+}
+func StartMonitoringService() {
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:142]: this should be go routine

File: server/pkg/client/vault2.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+		fmt.Println("Error while connecting to client", err)
+	}
+
+	vault := Vault{
+		client: client,
+	}
+
+	status, err := vault.client.Sys().SealStatus()
+	if err != nil {
+		fmt.Println("Error while sealing", err)
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:158]: improper error message

File: server/pkg/client/vault2.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+		fmt.Printf("Secret '%s' found in namespace '%s'\n", secret.Name, secret.Namespace)
+		// Use the secret as needed
+	}
+	return values
+}
+func StartMonitoringService() {
+
+	client, err := api.NewClient(&api.Config{
+		Address: "http://127.0.0.1:8200",
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:145]: vault address cannot be hardcoded like this

File: server/pkg/client/vault2.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+	})
+
+	if err != nil {
+		fmt.Println("Error while connecting to client", err)
+	}
+
+	vault := Vault{
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:152]: why don't we use the client directly. Why are we assigning the client to vault?

File: server/pkg/client/vault2.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+		client: client,
+	}
+
+	status, err := vault.client.Sys().SealStatus()
+	if err != nil {
+		fmt.Println("Error while sealing", err)
+	}
+	for {
+		if !status.Initialized {
```
Comment (srikrishnabh1) [inline server/pkg/client/vault2.go:161]: Status will always be same value until you update.

**Review by srikrishnabh1 — CHANGES_REQUESTED** [review]
Please fix the review comments

File: server/pkg/client/vault2.go (hunk 10 above)
Comment (Shifna12Zarnaz) [reply server/pkg/client/vault2.go:142]: yas....I have called in main

**Review by neha01 — CHANGES_REQUESTED** [review]

File: server/cmd/server/main.go [hunk 15]
```diff
@@ -6,6 +6,8 @@ import (
 	"os/signal"
 	"syscall"
 
+	"github.com/kube-tarian/kad/server/pkg/client"
```
[the remark this answers is not in the document]
Comment (Shifna12Zarnaz) [reply server/cmd/server/main.go:9]: sry am not getting you

**Comment by Shifna12Zarnaz** [comment]
Added vault2.go

**Comment by Shifna12Zarnaz** [comment]
Added vault2.go with specified changes

File: server/cmd/server/main.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (neha01) [reply server/cmd/server/main.go:9]: when you goimport file you group native go modules and third party modules separately.
https://pkg.go.dev/golang.org/x/tools/cmd/goimports
Medium article: https://rgalus.medium.com/how-to-format-your-go-projects-with-gofmt-goimports-and-git-d1fa3c3723ac

**Comment by vramk23** [comment]
not needed

**Pull request closed without merging**